File: src/agents/sentiment.py
import json
from typing import Dict, Any, List
from datetime import datetime, timedelta
from concurrent.futures import ThreadPoolExecutor
import pandas as pd
import numpy as np
from langchain_core.messages import HumanMessage
from langchain_openai.chat_models import ChatOpenAI
from langchain_core.prompts import ChatPromptTemplate
from graph.state import AgentState, show_agent_reasoning
from utils.progress import progress
from tools.api import get_news_data, get_insider_trades
from data.models import NewsArticle

def sentiment_agent(state: AgentState):
    """Analyzes market sentiment and generates trading signals for multiple tickers."""
    data = state.get("data", {})
    end_date = data.get("end_date")
    tickers = data.get("tickers")
    llm = ChatOpenAI(model="gpt-4o")

    sentiment_analysis = {}

    for ticker in tickers:
        sentiment_analysis[ticker] = _analyze_ticker_sentiment(ticker, end_date, llm)
        progress.update_status("sentiment_agent", ticker, "Done")

    message = HumanMessage(
        content=json.dumps(sentiment_analysis),
        name="sentiment_agent",
    )

    if state["metadata"]["show_reasoning"]:
        show_agent_reasoning(sentiment_analysis, "Sentiment Analysis Agent")

    state["data"]["analyst_signals"]["sentiment_agent"] = sentiment_analysis

    return {
        "messages": [message],
        "data": data,
    }

def _analyze_ticker_sentiment(ticker: str, end_date: str, llm: ChatOpenAI) -> Dict[str, Any]:
    """Analyze sentiment for a single ticker."""
    progress.update_status("sentiment_agent", ticker, "Fetching insider trades")
    insider_trades = get_insider_trades(
        ticker=ticker,
        end_date=end_date,
        limit=1000,
    )

    if not insider_trades:
        progress.update_status("sentiment_agent", ticker, "Failed: No insider trades found")
        return {
            "signal": "neutral",
            "confidence": 0,
            "reasoning": "No insider trades found"
        }

    progress.update_status("sentiment_agent", ticker, "Analyzing sentiment")
    
    with ThreadPoolExecutor() as executor:
        # Start news API call
        news_future = executor.submit(
            get_news_data,
            ticker=ticker,
            start_date=(datetime.now() - timedelta(days=7)).strftime('%Y-%m-%d'),
            end_date=end_date,
            sort_by="relevancy"
        )
        
        # Process insider trades while waiting for news
        insider_result = _process_insider_trades(insider_trades)
        
        # Get news results
        news_data = news_future.result()
        news_result = _process_news_sentiment(news_data, llm)

    # Combine sentiments (60% insider, 40% news)
    weighted_insider = insider_result["confidence"] * 0.6
    weighted_news = news_result["confidence"] * 0.4

    if news_result["signal"] == insider_result["signal"]:
        overall_sentiment = news_result["signal"]
        overall_confidence = weighted_insider + weighted_news
    else:
        if weighted_insider > weighted_news:
            overall_sentiment = insider_result["signal"]
            overall_confidence = weighted_insider
        else:
            overall_sentiment = news_result["signal"]
            overall_confidence = weighted_news

    return {
        "signal": overall_sentiment,
        "confidence": round(overall_confidence * 100, 2),
        "reasoning": {
            "insider_sentiment": {
                "signal": insider_result["signal"],
                "confidence": round(insider_result["confidence"] * 100, 2),
                "bullish_trades": insider_result["bullish_trades"],
                "bearish_trades": insider_result["bearish_trades"]
            },
            "news_sentiment": {
                "signal": news_result["signal"],
                "confidence": round(news_result["confidence"] * 100, 2),
                "articles_analyzed": news_result["articles_analyzed"],
                "key_points": news_result["key_points"]
            }
        }
    }

def _analyze_article_sentiment(article: NewsArticle, llm: ChatOpenAI) -> Dict[str, Any]:
    """Analyze sentiment of a single article using GPT."""
    template = ChatPromptTemplate.from_messages([
        (
            "system",
            """You are a financial sentiment analyzer. Analyze the following news article about a company.
            Consider only factors that directly impact stock value:
            1. Revenue/profit implications
            2. Market competition impact
            3. Regulatory/legal effects
            4. Innovation/product developments
            
            Return ONLY a JSON object with exactly these fields:
            {{
                "sentiment": "bullish" | "bearish" | "neutral",
                "confidence": <float between 0 and 1>,
                "key_points": [<max 2 most important points>]
            }}
            
            Be concise and focus only on stock price impact."""
        ),
        (
            "human",
            "Title: {title}\nDescription: {description}"
        ),
    ])

    prompt = template.invoke({
        "title": article.title,
        "description": article.description or ""
    })

    try:
        result = llm.invoke(prompt)
        return json.loads(result.content)
    except (json.JSONDecodeError, Exception):
        return {
            "sentiment": "neutral",
            "confidence": 0.0,
            "key_points": []
        }

def _process_insider_trades(trades: List[Dict[str, Any]]) -> Dict[str, Any]:
    """Process insider trading sentiment using pandas."""
    transaction_shares = pd.Series([t.transaction_shares for t in trades]).dropna()
    bearish_condition = transaction_shares < 0
    signals = np.where(bearish_condition, "bearish", "bullish").tolist()
    
    bullish_signals = signals.count("bullish")
    bearish_signals = signals.count("bearish")
    total_signals = len(signals)
    
    if total_signals > 0:
        sentiment = "bullish" if bullish_signals > bearish_signals else "bearish" if bearish_signals > bullish_signals else "neutral"
        confidence = max(bullish_signals, bearish_signals) / total_signals
    else:
        sentiment = "neutral"
        confidence = 0.0

    return {
        "signal": sentiment,
        "confidence": confidence,
        "bullish_trades": bullish_signals,
        "bearish_trades": bearish_signals
    }

def _process_news_sentiment(news_data: List[NewsArticle], llm: ChatOpenAI) -> Dict[str, Any]:
    """Process news sentiment from articles."""
    SENTIMENT_VALUES = {
        "bullish": 1,
        "bearish": -1,
        "neutral": 0
    }
    
    article_analyses = []
    for article in news_data[:5]:
        analysis = _analyze_article_sentiment(article, llm)
        if analysis["confidence"] > 0:
            article_analyses.append({
                "title": article.title,
                "source": article.source.name, 
                "published": article.publishedAt,
                "analysis": analysis
            })

    if not article_analyses:
        return {
            "signal": "neutral",
            "confidence": 0.0,
            "articles_analyzed": 0,
            "key_points": []
        }

    sentiment_scores = []
    total_confidence = 0
    all_key_points = []
    
    for analysis in article_analyses:
        sentiment_value = SENTIMENT_VALUES[analysis["analysis"]["sentiment"]]
        
        confidence = analysis["analysis"]["confidence"]
        sentiment_scores.append(sentiment_value * confidence)
        total_confidence += confidence
        all_key_points.extend(analysis["analysis"]["key_points"])

    avg_sentiment = sum(sentiment_scores) / len(sentiment_scores)
    confidence = total_confidence / len(article_analyses)
    
    sentiment = (
        "bullish" if avg_sentiment > 0.2
        else "bearish" if avg_sentiment < -0.2
        else "neutral"
    )

    return {
        "signal": sentiment,
        "confidence": confidence,
        "articles_analyzed": len(article_analyses),
        "key_points": list(set(all_key_points))
    }

File: src/agents/sentiment.rs
// Source: src/agents/sentiment.py
//! Sibling to src/agents/sentiment.py
//! Analyzes market sentiment by aggregating insider transaction volumes and company news sentiment.

use anyhow::{Result, Context};
use crate::graph::state::AgentState;
use crate::tools::api::{get_insider_trades, get_company_news};

/// Performs insider trading and news sentiment analysis, updating agent state.
pub async fn sentiment_analyst_agent(state: &mut AgentState, agent_id: &str) -> Result<()> {
    println!("Running Sentiment Analyst Agent: {}", agent_id);

    let end_date = state.data.get("end_date")
        .and_then(|v| v.as_str())
        .context("Missing end_date in sentiment_analyst_agent")?;

    let tickers_json = state.data.get("tickers")
        .context("Missing tickers in sentiment_analyst_agent")?;
    let tickers: Vec<String> = serde_json::from_value(tickers_json.clone())?;

    let api_key = state.metadata.get("FINANCIAL_DATASETS_API_KEY")
        .and_then(|v| v.as_str());

    let mut sentiment_analysis = serde_json::Map::new();

    for ticker in &tickers {
        // Fetch insider trades
        let insider_trades = get_insider_trades(ticker, end_date, None, 1000, api_key)
            .await
            .unwrap_or_default();

        // Get the signals from the insider trades
        let transaction_shares: Vec<f64> = insider_trades
            .iter()
            .filter_map(|t| t.transaction_shares)
            .collect();

        let insider_signals: Vec<&str> = transaction_shares
            .iter()
            .map(|&shares| if shares < 0.0 { "bearish" } else { "bullish" })
            .collect();

        // Fetch company news
        let company_news = get_company_news(ticker, end_date, None, 100, api_key)
            .await
            .unwrap_or_default();

        // Get the sentiment from the company news
        let sentiments: Vec<String> = company_news
            .iter()
            .filter_map(|n| n.sentiment.clone())
            .collect();

        let news_signals: Vec<&str> = sentiments
            .iter()
            .map(|s| {
                if s == "negative" {
                    "bearish"
                } else if s == "positive" {
                    "bullish"
                } else {
                    "neutral"
                }
            })
            .collect();

        // Combine signals from both sources with weights
        let insider_weight = 0.3;
        let news_weight = 0.7;

        let insider_bullish_count = insider_signals.iter().filter(|&&s| s == "bullish").count() as f64;
        let insider_bearish_count = insider_signals.iter().filter(|&&s| s == "bearish").count() as f64;

        let news_bullish_count = news_signals.iter().filter(|&&s| s == "bullish").count() as f64;
        let news_bearish_count = news_signals.iter().filter(|&&s| s == "bearish").count() as f64;
        let news_neutral_count = news_signals.iter().filter(|&&s| s == "neutral").count() as f64;

        // Calculate weighted signal counts
        let bullish_signals = insider_bullish_count * insider_weight + news_bullish_count * news_weight;
        let bearish_signals = insider_bearish_count * insider_weight + news_bearish_count * news_weight;

        let overall_signal = if bullish_signals > bearish_signals {
            "bullish"
        } else if bearish_signals > bullish_signals {
            "bearish"
        } else {
            "neutral"
        };

        // Calculate confidence level based on the weighted proportion
        let total_weighted_signals = (insider_signals.len() as f64) * insider_weight + (news_signals.len() as f64) * news_weight;
        let mut confidence = 0.0;
        if total_weighted_signals > 0.0 {
            let max_signals = if bullish_signals > bearish_signals { bullish_signals } else { bearish_signals };
            confidence = ((max_signals / total_weighted_signals) * 100.0 * 100.0).round() / 100.0;
        }

        let insider_signal = if insider_bullish_count > insider_bearish_count {
            "bullish"
        } else if insider_bearish_count > insider_bullish_count {
            "bearish"
        } else {
            "neutral"
        };

        let max_insider_count = if insider_bullish_count > insider_bearish_count { insider_bullish_count } else { insider_bearish_count };
        let insider_confidence = if !insider_signals.is_empty() {
            ((max_insider_count / insider_signals.len() as f64) * 100.0).round() as u64
        } else {
            0
        };

        let news_signal = if news_bullish_count > news_bearish_count {
            "bullish"
        } else if news_bearish_count > news_bullish_count {
            "bearish"
        } else {
            "neutral"
        };

        let max_news_count = if news_bullish_count > news_bearish_count { news_bullish_count } else { news_bearish_count };
        let news_confidence = if !news_signals.is_empty() {
            ((max_news_count / news_signals.len() as f64) * 100.0).round() as u64
        } else {
            0
        };

        let reasoning = serde_json::json!({
            "insider_trading": {
                "signal": insider_signal,
                "confidence": insider_confidence,
                "metrics": {
                    "total_trades": insider_signals.len(),
                    "bullish_trades": insider_bullish_count,
                    "bearish_trades": insider_bearish_count,
                    "weight": insider_weight,
                    "weighted_bullish": ((insider_bullish_count * insider_weight) * 10.0).round() / 10.0,
                    "weighted_bearish": ((insider_bearish_count * insider_weight) * 10.0).round() / 10.0,
                }
            },
            "news_sentiment": {
                "signal": news_signal,
                "confidence": news_confidence,
                "metrics": {
                    "total_articles": news_signals.len(),
                    "bullish_articles": news_bullish_count,
                    "bearish_articles": news_bearish_count,
                    "neutral_articles": news_neutral_count,
                    "weight": news_weight,
                    "weighted_bullish": ((news_bullish_count * news_weight) * 10.0).round() / 10.0,
                    "weighted_bearish": ((news_bearish_count * news_weight) * 10.0).round() / 10.0,
                }
            },
            "combined_analysis": {
                "total_weighted_bullish": (bullish_signals * 10.0).round() / 10.0,
                "total_weighted_bearish": (bearish_signals * 10.0).round() / 10.0,
                "signal_determination": format!(
                    "{} based on weighted signal comparison",
                    if bullish_signals > bearish_signals { "Bullish" } else if bearish_signals > bullish_signals { "Bearish" } else { "Neutral" }
                )
            }
        });

        sentiment_analysis.insert(
            ticker.clone(),
            serde_json::json!({
                "signal": overall_signal,
                "confidence": confidence,
                "reasoning": reasoning,
            }),
        );
    }

    // Print the reasoning if show_reasoning is enabled
    let show_reasoning = state.metadata.get("show_reasoning")
        .and_then(|v| v.as_bool())
        .unwrap_or(false);

    if show_reasoning {
        crate::graph::state::show_agent_reasoning(&serde_json::json!(sentiment_analysis), "Sentiment Analysis Agent");
    }

    // Add sentiment analysis to state's analyst signals
    let analyst_signals = state.data.entry("analyst_signals".to_string())
        .or_insert_with(|| serde_json::json!({}));
    if let Some(obj) = analyst_signals.as_object_mut() {
        obj.insert(agent_id.to_string(), serde_json::Value::Object(sentiment_analysis.clone()));
    }

    // Add a HumanMessage representing the aggregate analysis to the state messages list
    let message = serde_json::json!({
        "content": serde_json::to_string(&sentiment_analysis)?,
        "name": agent_id,
        "type": "human",
    });
    state.messages.push(message);

    Ok(())
}

